icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp

```diff
@@ -9,39 +16,404 @@ namespace icicle {
+    // TODO: handle size
+    eIcicleError build(const std::byte* leaves, uint64_t leaves_size, const MerkleTreeConfig& config) override
     {
-      ICICLE_LOG_DEBUG << "CPU CPUMerkleTreeBackend::build() called with " << size << " bytes of leaves";
-      return eIcicleError::SUCCESS; // TODO: Implement tree-building logic
+      TasksManager<HashTask> task_manager(get_nof_workers(config)); // Run workers.
+      ICICLE_ASSERT(!m_tree_already_built) << "Tree cannot be built more than one time";
+      const int nof_layers = m_layers.size();
+      m_tree_already_built = true; // Set the tree status as built
+      uint64_t l0_segment_idx = 0; // Index for the chunk of hashes from layer 0 to send
+      init_layers_db(config);
+      const uint64_t nof_segments_at_l0 =
+        (m_layers[0].m_nof_hashes + NOF_OPERATIONS_PER_TASK - 1) / NOF_OPERATIONS_PER_TASK;
+
+      // run until the root is processed
+      while (1) {
+        HashTask* task = (l0_segment_idx < nof_segments_at_l0) ? // If there are tasks from layer 0 to send
+                           task_manager.get_idle_or_completed_task()
+                                                               : // get any task slot to assign
+                           task_manager.get_completed_task();    // else, only completed tasks are interesting.
+
+        // handle completed task
+        if (task->is_completed()) {
+          if (task->m_layer_idx == nof_layers - 1) { // Root processed
+            // print_tree(leaves, leaves_size);
+            return eIcicleError::SUCCESS;
+          }
+          const uint64_t completed_layer_idx = task->m_layer_idx;
+          const uint64_t completed_segment_idx = task->m_segment_idx;
+
+          // delete completed_segment_id from the map
+          const uint64_t completed_segment_id = completed_segment_idx ^ (completed_layer_idx << 56);
+          m_map_segment_id_2_inputs.erase(completed_segment_id);
+
+          // Calculate Current-Segment-ID. The completed task generated inputs for Current-Segment
+          const uint64_t cur_layer_idx = completed_layer_idx + 1;
+          const uint64_t cur_segment_idx = task->m_next_segment_idx;
+          const uint64_t cur_segment_id = cur_segment_idx ^ (cur_layer_idx << 56);
+
+          // update m_map_segment_id_2_inputs with the data that is ready for process
+          auto cur_segment_iter = m_map_segment_id_2_inputs.find(cur_segment_id);
+          cur_segment_iter->second->m_nof_inputs_ready +=
+            m_layers[completed_layer_idx].m_hash.output_size() * NOF_OPERATIONS_PER_TASK;
+
+          // check if cur segment is ready to be executed
+          const Hash cur_hash = m_layers[cur_layer_idx].m_hash;
+          const uint64_t nof_hashes_in_seg = std::min(
+            m_layers[cur_layer_idx].m_nof_hashes - cur_segment_idx * NOF_OPERATIONS_PER_TASK,
+            uint64_t(NOF_OPERATIONS_PER_TASK));
+          ICICLE_ASSERT(nof_hashes_in_seg > 0) << "Edge case negative number of hashes";
```

Comment:
what is this edge case? bug? input is wrong? for the latter, can you identify it earlier and return a clear error?

Reply:
Koren added that.
Removed until full verification once he'll be back